drivers/iio/dac/ad5791.c

```diff
@@ -317,12 +397,103 @@ static int ad5791_write_raw(struct iio_dev *indio_dev,
 		val <<= chan->scan_type.shift;
 
 		return ad5791_spi_write(st, chan->address, val);
-
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return ad5791_set_sampling_freq(indio_dev, val);
 	default:
 		return -EINVAL;
 	}
 }
 
+static int ad5791_buffer_preenable(struct iio_dev *indio_dev)
+{
+	struct ad5791_state *st = iio_priv(indio_dev);
+	int ret;
+
+	if (st->pwr_down)
+		return -EINVAL;
+
+	ret =  pwm_enable(st->cnv_trigger);
+	if (ret)
+		return ret;
+
+	spi_bus_lock(st->spi->master);
+	spi_engine_ex_offload_enable(st->spi, true);
+
+	return 0;
+}
+
+static int ad5791_buffer_postdisable(struct iio_dev *indio_dev)
+{
+	struct ad5791_state *st = iio_priv(indio_dev);
+
+	spi_engine_ex_offload_enable(st->spi, false);
+	spi_bus_unlock(st->spi->master);
+	pwm_disable(st->cnv_trigger);
+
+	return 0;
+}
+
+static const struct iio_buffer_setup_ops ad5791_buffer_setup_ops = {
+	.preenable = &ad5791_buffer_preenable,
+	.postdisable = &ad5791_buffer_postdisable,
+};
+
+static void ad5791_pwm_diasble(void *data)
+{
+	pwm_disable(data);
+}
+
+static int ad5791_pwm_setup(struct spi_device *spi, struct ad5791_state *st)
+{
+	struct clk *ref_clk;
+	int ret;
+
+	ref_clk = devm_clk_get_enabled(&spi->dev, "ref_clk");
+	if (IS_ERR(ref_clk))
+		return PTR_ERR(ref_clk);
+
+	st->ref_clk_rate = clk_get_rate(ref_clk);
+
+	st->cnv_trigger = devm_pwm_get(&spi->dev, "cnv");
+	if (IS_ERR(st->cnv_trigger))
+		return PTR_ERR(st->cnv_trigger);
+
+	ret = devm_add_action_or_reset(&spi->dev, ad5791_pwm_diasble,
+				       st->cnv_trigger);
+	if (ret)
+		return ret;
+
+	return __ad5791_set_sampling_freq(st, AD5791_MAX_SAMPLING_RATE);
+}
+
+static int ad5791_offload_setup(struct iio_dev *indio_dev)
+{
+	struct ad5791_state *st = iio_priv(indio_dev);
+	int ret;
+
+	st->spi_transfer.len = 4;
+	st->spi_transfer.bits_per_word = 24;
+	st->spi_transfer.tx_buf = (void *)-1; /* steaming tx */
+
+	spi_message_init_with_transfers(&st->spi_msg, &st->spi_transfer, 1);
+
+	ret = devm_spi_optimize_message(&st->spi->dev, st->spi, &st->spi_msg);
+	if (ret)
+		return ret;
+
+	ret = spi_engine_ex_offload_load_msg(st->spi, &st->spi_msg);
+	if (ret < 0)
+		return ret;
+
+	ret = ad5791_pwm_setup(st->spi, st);
+	if (ret)
+		return ret;
+
+	return devm_iio_dmaengine_buffer_setup(indio_dev->dev.parent,
```

Comment:
```suggestion
	return devm_iio_dmaengine_buffer_setup(&st->spi->dev,
```

Technically the same device, but consistent with `devm_spi_optimize_message()` above.

Reply:
ok